### PR TITLE
Updates for Julia v0.7/v1.0 Compatibility

### PR DESCRIPTION
**Pull request by ccoffrin**
Most updates are fairly strait forward.  

The only non-trivial one is `findnz`.  It seems the return type semantics have changed which Compat not a viable option for this function.  My solution was to implement the function in this package using the recommended v1.0 revision and only add this function when running v0.7 or higher.

### CODE REVIEW

**Comment by chriscoey** [comment]
am I alone in thinking v0.6 support should be dropped?

**Comment by ccoffrin** [comment]
For the LANL packages I am endeavoring to make at least one version of every thing that simultaneously supports v0.6/v0.7/v1.0.  When that is done, everyone at LANL will have a path to update to v1.0 and we will drop v0.6 support in the subsequent version.

**Comment by chriscoey** [comment]
I'm more of a just rip the band-aid off kinda guy

**Comment by ccoffrin** [comment]
Right now MI* support is so spotty in v0.7, ripping of the band-aid would probably kill LANL. :-)

**Comment by chriscoey** [comment]
Right, but older releases will continue to support 0.6
Edit: major swypo

**Comment by chriscoey** [comment]
Fail autocorrect

**Comment by ccoffrin** [comment]
You have much more faith in the Julia packager manager than I.

**Comment by ccoffrin** [comment]
So far it's been really helpful for me that the solvers, e.g. SCS are working in all three.  It helps track down if new errors are due to the solver version update or the Julia version update.

**Comment by chriscoey** [comment]
I struggle to understand this. Julia 0.6 and oldpkg keep packages under .julia/v0.6, whereas Julia 0.7+ and pkg keep packages in a different location. 

I am prepared to go and update Pajarito soon but I think it's much easier and cleaner to just tag master with the right Julia version bounds and then start making 0.7+ only compatible changes and then tag again with updated version bounds.

**Comment by chriscoey** [comment]
If the upgrade is purely changing syntax for 0.7 then any regressions are due to a Julia version update.

There isn't really anything algorithmic that will change in Pavito and Pajarito.

**Comment by ccoffrin** [comment]
If you would like a tutorial in Julia v0.7 package development, I would be happy to provide one.  It is very different from Julia v0.6, but it is possible to support both simultaneously from one repo.

As long as,
1. your tests are comprehensive 
2. there are no unfinished new features in branches
3. all of your up-steam dependencies support v0.7
Then a hard jump from v0.6 to v0.7 seems reasonable.  

Basically all of the packages I work on fail point 2 (incomplete migrations to MOI) and 3 (limited MI solvers in v0.7)

I will note that using Compat, I found it pretty easy to support v0.6 using v0.7 features.

**Comment by chriscoey** [comment]
Ok. I suppose part of my reluctance to support both on master stems from (a) desire for simplicity of the code and (b) belief that we should not make it easy for people to just stay on 0.6 once the package supports 1.0.

Anyway, so this suggests a division of labor. It sounds like updating Pajarito for MOI is a much bigger job than updating for 1.0. If you would care to do the 1.0 update (keeping v0.6 support if you need) then I'll subsequently update for MOI. That will essentially be a rewrite of half of Pajarito's codebase. Pavito should be easier because the NLP interface is mostly unchanged.

**Comment by chriscoey** [comment]
Pajarito depends on ConicBenchmarkUtilities which is also not updated to 1.0 so that will also block the Pajarito update.

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,14 @@
...
 os:
   - linux
-#  - osx # disabled until queues are reasonable
+  - osx
 julia:
   - 0.6
   - 0.7
-#  - nightly
+  - 1.0
+  - nightly
```
Comment (mlubin) [inline .travis.yml]: Please drop nightly. It slows CI times for all JuliaOpt packages because of the shared queues.

File: Project.toml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+name = "ConicBenchmarkUtilities"
```
Comment (mlubin) [inline Project.toml:1]: Last time I checked project.toml files and manifest.toml files are not needed for registered packages and can confuse the package manager (https://github.com/JuliaRegistries/General/issues/22). Is there more recent guidance?

File: src/ConicBenchmarkUtilities.jl [hunk 3]
```diff
@@ -4,6 +4,21 @@ module ConicBenchmarkUtilities
...
+if VERSION < v"0.7.0-"
+    import Compat: undef
+    import Compat: @warn
+end
+
+if VERSION > v"0.7.0-"
+    using SparseArrays
+    using LinearAlgebra
+    # this is required because findall return type changed in v0.7
+    function SparseArrays.findnz(A::AbstractMatrix)
```
Comment (mlubin) [inline src/ConicBenchmarkUtilities.jl:16]: It's better to update the code to work with the new return types so that later we can just delete the 0.6 compatibility blocks and have clean 0.7/1.0 code.

File: test/runtests.jl [hunk 4]
```diff
@@ -25,17 +36,20 @@ using ConicBenchmarkUtilities
     x_sol = MathProgBase.getsolution(m)
     objval = MathProgBase.getobjval(m)
 
+    #=
+    # drop Convex.jl
```
Comment (mlubin) [inline test/runtests.jl]: Better to just delete the code instead of leaving a commented block that's unlikely to be revived. The code is stored in the git history if we need it.

**Review by mlubin — COMMENTED** [review]
Thanks for updating! A few comments.

File: .travis.yml (hunk 1 above)
Comment (ccoffrin) [reply .travis.yml]: Done

File: test/runtests.jl (hunk 4 above)
Comment (ccoffrin) [reply test/runtests.jl]: Done

File: Project.toml (hunk 2 above)
Comment (ccoffrin) [reply Project.toml:1]: It would seem that I do not know enough about the new package manager to understand how to use it without toml files.  For example, how does one run `test` inside the Pkg REPL?  If there is a tutorial on how to do this I can look into dropping the toml files.

For what it is worth, when setting up the toml files I use the same UUID as the General registry.  So I don't think there will be any confusion.

File: src/ConicBenchmarkUtilities.jl (hunk 3 above)
Comment (ccoffrin) [reply src/ConicBenchmarkUtilities.jl:16]: Agreed this would be preferable.  I tried to do it that way for about an hour, but in the end I think it is impossible or requires a much deeper knowledge of sparse matrices in Julia than I have.

The root of the challenge is that the semantics of `findall` have also changed in v0.7.  For the tests in this package, the v0.6 `findall` returns a list of integers (not clear how ints map back to matrix cells) and the v0.7 `findall` returns a list of Cartesian data types.  So the workaround code proposed for v0.7 is incompatible with v0.6.  Using Compat on `findall` to bring the v0.7 semantics into v0.6 is also no successful.

**Comment by ccoffrin** [comment]
@chriscoey, I'll make you a counter offer.

You write some new tests for `ConicNonlinearBridge`, so that the test code coverage is above 80% and I will make `Pavito` and `Pajarito` work with JuMP v0.18 in Julia v0.6/0.7/1.0.  After that is done, migrations to MOI and JuMP v0.19 can be discussed. :-)

**Comment by chriscoey** [comment]
Ok deal. Let me know when to start on tests!

**Comment by ccoffrin** [comment]
Add JuMP-based tests to `https://github.com/mlubin/ConicNonlinearBridge.jl` ASAP.  Once that done, I should be able to update it in short order.

**Comment by ccoffrin** [comment]
I need those tests before I can start migrating `Pavito`.

**Comment by chriscoey** [comment]
old JuMP/MPB doesn't really support cone modeling, so I used MPB directly. see PR at https://github.com/mlubin/ConicNonlinearBridge.jl/pull/10

**Comment by mlubin** [comment]
@ccoffrin I moved the repo to JuliaOpt and gave you access. Tag when ready.

**Comment by ccoffrin** [comment]
Thanks!!!  I am gonna play around with dropping the toml files today.  I'll tag after I have some more clarity on that point.
